Rework readme (#68)
Sensirion/embedded-common: README.md

# embedded-common
This repository contains the implementation of the Sensirion I2C and SHDLC
protocol and some commonly used functions for sensor implementations.
It is a dependency of the Sensirion embedded sensor drivers. Sensirion
senors all use a protocol on top of the I2C/UART communications layer.

> There are two of those kinds protocols used by Sensirion sensors, one

Reviewer: ```suggestion
There are two of those kinds of protocols used by Sensirion sensors, one
```

> senors all use a protocol on top of the bit level communications like I2C or
> UART. There are two of those kinds protocols used by Sensirion sensors, one
> for UART communication and one for I2C communication. These two work the same
> for all sensor types and therefore need to be implemented only once. To keep

Reviewer: (please reformat when applying)

```suggestion
senors all use a protocol on top of the I2C/UART communications layer.
There are two of those kinds protocols used by Sensirion sensors, one
for UART communication, called SHDLC, and one for I2C communication.
These are shared amongst sensor types and are therefore implemented only once.
To keep
```

> This repository contains the implementation of the Sensirion I2C and SHDLC
> protocol and some commonly used functions for sensor implementations.
> It is a dependency of the Sensirion embedded sensor drivers. Sensirion
> senors all use a protocol on top of the bit level communications like I2C or
> UART. There are two of those kinds protocols used by Sensirion sensors, one
> for UART communication and one for I2C communication. These two work the same
> for all sensor types and therefore need to be implemented only once. To keep
> this implementation in one place this repository exists and is used by all
> embedded drivers.

Author: ```suggestion
senors all use a protocol on top of the I2C/UART communications layer.
There are two of those kinds protocols used by Sensirion sensors, one
for UART communication, called SHDLC, and one for I2C communication.
These are shared amongst sensor types and are therefore implemented only once. 
To keep this implementation in one place this repository exists and is used by all
embedded drivers.
```

> senors all use a protocol on top of the bit level communications like I2C or
> UART. There are two of those kinds protocols used by Sensirion sensors, one
> for UART communication and one for I2C communication. These two work the same
> for all sensor types and therefore need to be implemented only once. To keep
> this implementation in one place this repository exists and is used by all
> embedded drivers.

Author: ```suggestion
senors all use a protocol on top of the I2C/UART communications layer.
There are two of those kinds protocols used by Sensirion sensors, one
for UART communication, called SHDLC, and one for I2C communication.
These are shared amongst sensor types and are therefore implemented only once. 
To keep this implementation in one place this repository exists and is used by all
embedded drivers.
```

This repository contains the implementation of the Sensirion I2C and SHDLC
protocol and some commonly used functions for sensor implementations.
It is a dependency of the Sensirion embedded sensor drivers. Sensirion
sensors all use a protocol on top of the I2C/UART communications layer.
There are two of those kinds of protocols used by Sensirion sensors, one
for UART communication, called SHDLC, and one for I2C communication.

> These are shared amongst sensor types and are therefore implemented only once.

Reviewer 2: no need for second are... suggestion leave it away completely

> To keep this implementation in one place this repository exists and is used by all

Reviewer 2: ... in one place only, this repository ...

https://github.com/Sensirion/info#repositories

## Repository content

> The source code in this repository can be spilt into three parts.

Reviewer: ```suggestion
The source code in this repository is spilt into three parts: Common code, I2C protocol and SHDLC protocol.
```

### Common

> In the `common/` folder are three files. `senirion_config.h` is a general

Reviewer: ```suggestion
In the `common/` folder are three files. `senirion_config.h` is the general
```

configuration file. Depending on your hardware you will need to edit this.

> You can find more about this further down. There are also

Reviewer: ```suggestion
For more information, see the section named `sensirion_config.h` further down in this document. There are also
```

`sensirion_common.[ch]`. These contain common helper functions used by both
UART and I2C sensor drivers.

### I2C

In the `i2c/` folder is the implementation of the protocol used by Sensirion
sensors for I2C communication. This implementation is spilt into two parts.
The implementation of the high level protocol in the files `sensirion_i2c.[ch]`
and a hardware abstraction layer in `senirion_i2c_hal.[ch]`.

> In `sensirion_i2c.[ch]` all the implementation details (CRC calculation,byte order,

Reviewer: ```suggestion
In `sensirion_i2c.[ch]` all the implementation details (CRC calculation, byte order swapping,
```

...) of Sensirion's I2C communication are implemented. The finished frame is then

> send to your hardware using the HAL (Hardware Abstraction Layer) in the other

Reviewer: ```suggestion
sent to the attached hardware using the Hardware Abstraction Layer (HAL) in the other
```

two files. Since the way how to send out this frame depends on underlying

> hardware and operating system we provide different implementations for that part

Reviewer: ```suggestion
hardware and operating system, we provide different implementations for that part
```

of the code. You can find these in the folder `i2c/sample-implementations/` for the
most common platforms. We even provide a implementation for software I2C using
bit-banging on GPIOs in `i2c/sample-implementations/GPIO_bit_banging/`. If your

> platform isn't there you can use the empty structure in the `i2c/` folder to

Reviewer: ```suggestion
platform isn't currently there, you can use the provided template structure in the `i2c/` folder to
```

> implement it yourself.

Reviewer: ```suggestion
implement it yourself. We're very happy to review and include more architectures in the form of a pull request on GitHub.
```

### SHDLC

In the `shdlc` folder is the implementation of the protocol used by Sensirion
sensors for UART communication. These sensors use Sensirions own SHDLC protocol.
The high level implementation of this protocol can be found in
`sensirion_shdlc.[ch]` and the hardware abstraction layer in `sensirion_uart_hal.[ch]`.
In `sensirion_shdlc.[ch]` all the implementation details (checksum calculation,

> byte order, frame constructon, ...) of Sensirion's SHDLC communication are

Reviewer: ```suggestion
byte order swapping, frame construction, ...) of Sensirion's SHDLC communication are
```

> implemented. The finished frame is then send to your hardware using the HAL

Reviewer: ```suggestion
implemented. The finished frame is then sent to your hardware using the HAL
```

> (Hardware Abstraction Layer) in the other two files. Since the way how to send

Reviewer: ```suggestion
in the other two files. Since the way how to send
```

out this frame depends on underlying hardware and operating system we provide
different implementations for that part of the code. You can find these in the
folder `shdlc/sample-implementaions/` for the most common platforms. If your

> platform isn't there you can use the empty structure in the `shdlc/` folder to

Reviewer: ```suggestion
platform isn't currently there, you can use the provided template structure in the `shdlc/` folder to
```

> implement it yourself.

Reviewer: ```suggestion
implement it yourself. We're very happy to review and include more architectures in the form of a pull request on GitHub.
```

## Files to be adjusted by the User

The following files are the only ones you should need to change if you want to
use the functionality of `embedded-common`. Here we explain which changes need
to be made according to your hardware platform.

### sensirion\_config.h

In this file we keep all the included libraries for our drivers. It is located

> in the `common/` folder.  The two libraries our drivers use are `<stdint.h>` and

Reviewer: ```suggestion
in the `common/` folder. The two libraries our drivers use are `<stdint.h>` and
```

`<stdlib.h>`. The features we use out of those libraries is standard integer sizes

> from `<stdint.h>` and NULL from `<stdlib.h>`. We assume you have both those libraries

Reviewer: ```suggestion
from `<stdint.h>` and `NULL` from `<stdlib.h>`. We assume that those libraries are provided by your toolchain, compiler or are
```

sensors for I2C communication. This implementation is spilt into two parts.
The implementation of the high level protocol in the files `sensirion_i2c.[ch]`
and a hardware abstraction layer in `senirion_i2c_hal.[ch]`.
In `sensirion_i2c.[ch]` all the implementation details (CRC calculation, byte order
swapping, ...) of Sensirion's I2C communication are implemented. The finished frame
is then sent to the attached hardware using the Hardware Abstraction Layer (HAL) in
the other two files. Since the way how to send out this frame depends on underlying
hardware and operating system, we provide different implementations for that part
of the code. You can find these in the folder `i2c/sample-implementations/` for the
most common platforms. We even provide a implementation for software I2C using
bit-banging on GPIOs in `i2c/sample-implementations/GPIO_bit_banging/`. If your
platform isn't currently there, you can use the provided template structure in the
`i2c/` folder to implement it yourself. We're very happy to review and include more
architectures in the form of a pull request on GitHub.

### SHDLC

In the `shdlc` folder is the implementation of the protocol used by Sensirion

> sensors for UART communication. These sensors use Sensirions own SHDLC protocol.

Reviewer 3: ```suggestion
sensors for UART communication. These sensors use Sensirion's own SHDLC protocol.
```

You use Sensirion's everywhere else. I'd also be fine with using Sensirions everywhere.

## Repository content

The source code in this repository is spilt into three parts: Common code,
I2C protocol and SHDLC protocol.

### Common

In the `common/` folder are three files. `senirion_config.h` is the general
configuration file. Depending on your hardware you will need to edit this.
For more information, see the section named `sensirion_config.h` further down

> in this document. There are also `sensirion_common.[ch]`. These contain common

Reviewer 2: There are also the two files ...

helper functions used by both UART and I2C sensor drivers.

### I2C

In the `i2c/` folder is the implementation of the protocol used by Sensirion
sensors for I2C communication. This implementation is spilt into two parts.
The implementation of the high level protocol in the files `sensirion_i2c.[ch]`
and a hardware abstraction layer in `senirion_i2c_hal.[ch]`.
In `sensirion_i2c.[ch]` all the implementation details (CRC calculation, byte order
swapping, ...) of Sensirion's I2C communication are implemented. The finished frame
is then sent to the attached hardware using the Hardware Abstraction Layer (HAL) in
the other two files. Since the way how to send out this frame depends on underlying
hardware and operating system, we provide different implementations for that part
of the code. You can find these in the folder `i2c/sample-implementations/` for the
most common platforms. We even provide a implementation for software I2C using
bit-banging on GPIOs in `i2c/sample-implementations/GPIO_bit_banging/`. If your

> platform isn't currently there, you can use the provided template structure in the

Reviewer 2: If we don't provide code for your platform yet you can use ...

`i2c/` folder to implement it yourself. We're very happy to review and include more
architectures in the form of a pull request on GitHub.

### SHDLC

> In the `shdlc` folder is the implementation of the protocol used by Sensirion

Reviewer 2: The shdlc folder contains ...

sensors for UART communication. These sensors use Sensirion's own SHDLC protocol.
The high level implementation of this protocol can be found in
`sensirion_shdlc.[ch]` and the hardware abstraction layer in `sensirion_uart_hal.[ch]`.
In `sensirion_shdlc.[ch]` all the implementation details (checksum calculation,
byte order swapping, frame construction, ...) of Sensirion's SHDLC communication are
implemented. The finished frame is then sent to your hardware using the HAL
in the other two files. Since the way how to send
out this frame depends on underlying hardware and operating system we provide
different implementations for that part of the code. You can find these in the
folder `shdlc/sample-implementaions/` for the most common platforms. If your

> platform isn't currently there, you can use the provided template structure in the

Reviewer 2: see above

* `int8_t` = signed 8bit integer
* `uint8_t` = unsigned 8bit integer

In addition to that you will need to specify `NULL`. For both we have a commented
template where just need to fill in your system specific values.

Make sure you specify the whole range (64bit sized integers are used on a 32bit
system for some sensor data).

### sensirion\_i2c\_hal.c

As already explained earlier is all the I2C related code in this repository based

> on Sensirion's I2c HAL. It is defined in `sensirion_i2c_hal.h` and implemented in

Reviewer: ```suggestion
on Sensirion's I2C HAL. It is defined in `sensirion_i2c_hal.h` and implemented in
```

* `int8_t` = signed 8bit integer
* `uint8_t` = unsigned 8bit integer

In addition to that you will need to specify `NULL`. For both we have a commented
template where just need to fill in your system specific values.

Make sure you specify the whole range (64bit sized integers are used on a 32bit
system for some sensor data).

### sensirion\_i2c\_hal.c

As already explained earlier is all the I2C related code in this repository based
on Sensirion's I2C HAL. It is defined in `sensirion_i2c_hal.h` and implemented in

> the corresponding `.c` file, both located in the `i2c/` folder. To use the I2c part

Reviewer 3: ```suggestion
the corresponding `.c` file, both located in the `i2c/` folder. To use the I2C part
```

of `embedded-common` you either need to implement the `sensirion_i2c_hal.c` file or
replace the unimplemented version in `i2c/` with the correct sample implementation

> from `i2c/sample-implementation/`.

Reviewer 3: I found this sentence a bit confusing: Maybe better: "you either need to implement the functions in `sensirion_i2c_hal.c` or replace it with a sample implementation from `i2c/sample-implementation/`".

> of `embedded-common` you either need to implement the `sensirion_i2c_hal.c` file or
> replace the unimplemented version in `i2c/` with the correct sample implementation
> from `i2c/sample-implementation/`.

Reviewer 3: ```suggestion
of `embedded-common` you either need to implement the functions in sensirion_i2c_hal.c or
replace it with a sample implementation from i2c/sample-implementation/.
```
